pkg/controller/batch/batch_controller.go

```diff
@@ -0,0 +1,373 @@
+}
+
+func (r *batchResource) validate() (reconcile.Result, error) {
+	spec := r.instance.Spec
+
+	if spec.ConfigMap == nil && spec.Config == nil {
+		return reconcile.Result{},
+			r.UpdatePhase(v2.BatchFailed, fmt.Errorf("'Spec.config' OR 'spec.ConfigMap' must be configured"))
+	}
+
+	if spec.ConfigMap != nil && spec.Config != nil {
+		return reconcile.Result{},
+			r.UpdatePhase(v2.BatchFailed, fmt.Errorf("At most one of ['Spec.config', 'spec.ConfigMap'] must be configured"))
+	}
+	return reconcile.Result{}, r.UpdatePhase(v2.BatchInitializing, nil)
+}
+
+func (r *batchResource) initializeResources() (reconcile.Result, error) {
+	batch := r.instance
+	spec := batch.Spec
+	// Ensure the Infinispan cluster exists
+	infinispan := &v1.Infinispan{}
+	if result, err := kube.LookupResource(spec.Cluster, batch.Namespace, infinispan, r.client, log); result != nil {
+		return *result, err
+	}
+
+	if err := infinispan.EnsureClusterStability(); err != nil {
+		log.Info(fmt.Sprintf("Infinispan '%s' not ready: %s", spec.Cluster, err.Error()))
+		return reconcile.Result{RequeueAfter: consts.DefaultWaitOnCluster}, nil
+	}
+
+	if spec.ConfigMap == nil {
+		// Create configMap
+		configMap := &corev1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      batch.Name,
+				Namespace: batch.Namespace,
+			},
+		}
+		_, err := controllerutil.CreateOrUpdate(ctx, r.client, configMap, func() error {
+			configMap.Data = map[string]string{BatchFilename: *spec.Config}
+			return controllerutil.SetControllerReference(batch, configMap, r.scheme)
+		})
+
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Unable to create ConfigMap '%s': %w", configMap.Name, err)
+		}
+
+		updated, err := r.update(func() error {
+			batch.Spec.ConfigMap = &batch.Name
+			return controllerutil.SetControllerReference(infinispan, batch, r.scheme)
+		})
+		if err != nil {
+			return reconcile.Result{}, fmt.Errorf("Unable to update Batch .Spec: %w", err)
+		}
+		if updated {
+			return reconcile.Result{}, nil
+		}
+	}
+
+	// We update the phase separately to the spec as the status update is ignored when in the update mutate function
+	return reconcile.Result{}, r.UpdatePhase(v2.BatchInitialized, nil)
+}
+
+func (r *batchResource) execute() (reconcile.Result, error) {
+	batch := r.instance
+	infinispan := &v1.Infinispan{}
+	if result, err := kube.LookupResource(batch.Spec.Cluster, batch.Namespace, infinispan, r.client, log); result != nil {
+		return *result, err
+	}
+
+	url, err := connectionUrl(infinispan)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	cliArgs := fmt.Sprintf("--connect '%s' --trustall --file '%s/%s'", url, BatchVolumeRoot, BatchFilename)
+
+	labels := batchLabels(batch.Name)
+	infinispan.AddLabelsForPods(labels)
+
+	job := &batchv1.Job{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      batch.Name,
+			Namespace: batch.Namespace,
+		},
+		Spec: batchv1.JobSpec{
+			BackoffLimit: pointer.Int32Ptr(0),
+			Template: corev1.PodTemplateSpec{
```

Comment:
@ryanemerson we also discussed that Job Pods should have a Batch CR as owner reference to prevent orphaned Pods after Batch CR deletion.

Reply:
As the pods are owned by the Job, I think the only way todo this would be to manually get the created pod and set the owner reference. Do you know of a better way todo this? For example via the Job podspec?

Reply:
Yeah, seems to be a problem with Job controller implementation on some old OCP version I've experimented before.

Reply:
Please see comment about cleanup resources. On latest OCP Job Pods are deleted with Job garbage collection